public/js/main.js

```diff
@@ -1,28 +1,15 @@
-import { _ } from './util.js';
-import { $search, $rollingKeywords, $recommendedKeyword } from './ref.js';
+import { _ } from './util/util.js';
+import { $searchBarInput, $rollingKeywords, $recommendedKeyword } from './util/ref.js';
+import { URL, KEYCODE } from './util/data.js';
+import initSearchBar from './searchBar.js';
 const { log } = console;
 
-$search.addEventListener('input', () => {
-  const searchInput = $search.value;
-  fetch(`
-  https://completion.amazon.com/api/2017/suggestions?session-id=146-7463946-6351015&customer-id=&request-id=GJ3692SXTPG2EG7P3KFS&page-type=Gateway&lop=en_US&site-variant=desktop&client-info=amazon-search-ui&mid=ATVPDKIKX0DER&alias=aps&b2b=0&fresh=0&ks=undefined&prefix=${searchInput}&event=onFocusWithSearchTerm&limit=11&fb=1&suggestion-type=KEYWORD&suggestion-type=WIDGET&_=1615392526147
-  `)
+$searchBarInput.addEventListener('input', () => {
+  const searchInput = $searchBarInput.value;
+  fetch(URL.autoComplete(searchInput))
     .then(res => res.json())
     .then(jsonData => jsonData.suggestions.map(v => v.value))
     .then(console.log);
 })
 
-const renderRecommendedKeywords = async () => {
-  const recommendedKeywords = await fetch(`
-    https://shoppinghow.kakao.com/v1.0/shophow/top/recomKeyword.json
-  `)
-    .then(res => res.json())
-    .then(jsonData => jsonData.list.map(v => v.keyword));
-
-  const renderedValue = recommendedKeywords.reduce((prev, curr, idx) => {
-    return prev + `<li>${recommendedKeywords[idx]}</li>`
-  }, '<ol>')
-  $recommendedKeyword.innerHTML = renderedValue + '</ol>';
-}
-
-renderRecommendedKeywords();
\ No newline at end of file
+initSearchBar();
```

Comment:
`searchBar.js`에서 드린 코멘트와 연결되는 부분이에요. 지금 구조와 제가 제안 드린 구조로 변경했을 때 장단점을 생각해 보셨으면 좋겠어요. 추후에 무언가 수정되거나 추가되어야 할 때, 어떤 방식이 더 수정 비용이 낮을까 고민해보시면 많은 도움이 될 것 같아요.
```suggestion
const searchBar = new SearchBar({ ui: $searchBar, inputArea: $searchBarInput, suggestions: $searchSuggestions });
searchBar.registerEvent();
const keywordSuggestion = new KeywordSuggestion({ $recommendedKeyword });
keywordSuggestion.render();
```

Reply:
너무 신기해요. 이렇게 쓸 수 있다니!!! 감사합니다 🥺🥺🥺
제가 현재는 `searchBar.js` 모듈에서 클래스와 메서드들을 정의하고,
그 모듈에서 init 함수를 만들고 실행이 필요한 함수들을 init에 담은 뒤,
init 함수만 export 해서 main에서는 init함수만 (import하면서 이름을 `initSearchBar`로 바꿈) 호출하도록 만들었는데 이 방법은 좋지 않은 방법인가요 ??

Reply:
안 좋은 방법은 아니라고 생각해요 😃 필요에 따라 initialize하는 코드가 여러 곳에서 중복으로 사용 된다면 지금의 `init`메서드가 충분히 의미 있겠죠. 다만 제가 `init`에서 하는 일을 main.js로 옮기는 것을 제안드린 이유는 main.js에서만 사용되고 있기도 하고, class를 활용하신다면 import, export하는 모듈 단위도 class로 하는게 좋다는 생각이 들어서에요. 그리고 마지막으로 코멘트 드린 것 처럼 `searchBar.js`에서 `ref.js`를 import하지 않아도 되기 때문이에요. 마지막 목적만 달성하기 위함이라면 `init` 메서드를 지금처럼 `searchBar.js`에 두고 아래와 같이 작성해도 괜찮긴 한데요, 앞서 말씀드린 다른 이유와 함께 생각해봤을 때 main.js로 옮기는 것이 좋겠다고 말씀드렸어요.

```javascript
const init = ({ $searchBar, $searchBarInput, $searchSuggestions, $recommendedKeyword }) => {
  const searchBar = new SearchBar({ $searchBar, $searchBarInput, $searchSuggestions });
  searchBar.registerEvent();
  const keywordSuggestion = new KeywordSuggestion($recommendedKeyword);
  keywordSuggestion.render();
};
```

Reply:
답변 정말 x 100000 감사드립니다!! 감동감동 🥺